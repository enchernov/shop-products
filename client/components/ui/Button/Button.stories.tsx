import React from 'react'
import { Story } from '@storybook/react'
import { Meta } from '@storybook/react/types-6-0'
import { action } from '@storybook/addon-actions'

import Button from './index'
import { IButtonProps } from './Button'

export default {
  title: 'ui/Button',
  component: Button,
  argTypes: {
    children: {
      control: {
        type: 'text',
      },
      defaultValue: 'Button',
    },
    disabled: {
      control: {
        type: 'boolean',
      },
    },
    icon: {
      control: {
        type: 'select',
        options: ['create', 'edit'],
      },
    },
  },
} as Meta

const Template: Story<IButtonProps> = (args) => (
  <Button onClick={action('clicked')} {...args} />
)

export const Text = Template.bind({})
Text.args = {
  children: 'Button',
  disabled: false,
}

export const Icon = Template.bind({})
Icon.args = {
  ...Text.args,
  icon: 'edit',
}

export const Link = Template.bind({})
Link.args = {
  ...Icon.args,
  href: '/',
}
