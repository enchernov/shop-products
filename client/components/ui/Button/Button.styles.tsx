import { createStyles, makeStyles } from '@material-ui/core/styles'

export const useStyles = makeStyles(() =>
  createStyles({
    button: {
      fontSize: '0.7rem',
      boxShadow: 'none',
      borderRadius: 0,
      '&:hover': {
        boxShadow: 'none',
      },
    },
  })
)
