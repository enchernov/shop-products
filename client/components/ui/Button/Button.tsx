import React, {
  FunctionComponent,
  ElementType,
  forwardRef,
  Ref,
  ReactNode,
} from 'react'
import {
  Button as MuiButton,
  ButtonProps,
  SvgIconProps,
} from '@material-ui/core'
import { Edit, AddSharp } from '@material-ui/icons'

import Link from '../Link'
import { useStyles } from './Button.styles'

type IconType = 'create' | 'edit'

export interface IButtonProps {
  children: ReactNode
  icon?: IconType
  disabled?: boolean
  href?: string
  className?: string
  fullwidth?: boolean
  onClick?: () => void
}

const defaultProps: Partial<IButtonProps> = {
  disabled: false,
}

type DefaultProps = Readonly<typeof defaultProps>

type ButtonPropsType = ButtonProps & IButtonProps & DefaultProps

const iconButtonMap: Record<IconType, ElementType<SvgIconProps>> = {
  create: AddSharp,
  edit: Edit,
}

const noop = () => {}

const Button: FunctionComponent<ButtonPropsType> = forwardRef(
  (
    { icon, children, disabled, href, onClick, fullWidth },
    ref: Ref<HTMLButtonElement>
  ) => {
    const classes = useStyles()

    const IconComponent: any = icon ? iconButtonMap[icon] : null

    const startIcon = icon ? { startIcon: <IconComponent /> } : undefined

    const linkProps = href ? { href: href, component: Link } : undefined

    return (
      <MuiButton
        ref={ref}
        color="primary"
        variant="contained"
        size="small"
        disabled={disabled}
        className={classes.button}
        {...startIcon}
        {...linkProps}
        fullWidth={fullWidth}
        onClick={!disabled ? onClick : noop}
      >
        {children}
      </MuiButton>
    )
  }
)

Button.defaultProps = defaultProps

export default Button
