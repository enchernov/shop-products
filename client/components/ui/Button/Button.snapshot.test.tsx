import React from 'react'
import Button from '../Button'
import renderer from 'react-test-renderer'

it('renders correctly', () => {
  const tree = renderer.create(<Button text="Some Text" />).toJSON()
  expect(tree).toMatchSnapshot()
})

// import { render, fireEvent } from '@testing-library/react';
// it('should respond to click events', () => {
//   const handleClick = jest.fn()
//   const instance = render(<Button {...Text.args} onClick={handleClick} />)
//   fireEvent.click(instance.container.firstChild)
//   expect(handleClick).toHaveBeenCalled()
// })
