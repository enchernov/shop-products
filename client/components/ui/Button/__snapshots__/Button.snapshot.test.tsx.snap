// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-1 MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    Some Text
  </span>
</button>
`;
